extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java

```diff
@@ -19,50 +19,88 @@
 import org.apache.druid.data.input.impl.CloudObjectLocation;
+import org.apache.druid.data.input.impl.S3ConfigProperties;
 import org.apache.druid.data.input.impl.SplittableInputSource;
 import org.apache.druid.storage.s3.S3InputDataConfig;
+import org.apache.druid.storage.s3.S3StorageConfig;
 import org.apache.druid.storage.s3.S3StorageDruidModule;
 import org.apache.druid.storage.s3.S3Utils;
 import org.apache.druid.storage.s3.ServerSideEncryptingAmazonS3;
 
 import javax.annotation.Nullable;
 import java.net.URI;
 import java.util.List;
+import java.util.Objects;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
 public class S3InputSource extends CloudObjectInputSource<S3Entity>
 {
-  private final ServerSideEncryptingAmazonS3 s3Client;
+  // We lazily initialize s3Client to avoid costly s3 operation when we only need S3InputSource for stored information
+  // (such as for task logs) and not for ingestion. (This cost only applies for new s3Client created with
+  // s3ConfigProperties given).
+  private final Supplier<ServerSideEncryptingAmazonS3> s3Client;
+  @JsonProperty("properties")
+  private final S3ConfigProperties s3ConfigProperties;
   private final S3InputDataConfig inputDataConfig;
 
   @JsonCreator
   public S3InputSource(
       @JacksonInject ServerSideEncryptingAmazonS3 s3Client,
+      @JacksonInject AmazonS3ClientBuilder amazonS3ClientBuilder,
+      @JacksonInject S3StorageConfig storageConfig,
       @JacksonInject S3InputDataConfig inputDataConfig,
       @JsonProperty("uris") @Nullable List<URI> uris,
       @JsonProperty("prefixes") @Nullable List<URI> prefixes,
-      @JsonProperty("objects") @Nullable List<CloudObjectLocation> objects
+      @JsonProperty("objects") @Nullable List<CloudObjectLocation> objects,
+      @JsonProperty("properties") @Nullable S3ConfigProperties s3ConfigProperties
   )
   {
     super(S3StorageDruidModule.SCHEME, uris, prefixes, objects);
-    this.s3Client = Preconditions.checkNotNull(s3Client, "s3Client");
     this.inputDataConfig = Preconditions.checkNotNull(inputDataConfig, "S3DataSegmentPusherConfig");
+    this.s3ConfigProperties = s3ConfigProperties;
+    this.s3Client = Suppliers.memoize(
+        () -> {
+          if (amazonS3ClientBuilder != null && storageConfig != null && s3ConfigProperties != null) {
+            if (s3ConfigProperties.isCredentialsConfigured()) {
+              BasicAWSCredentials creds = new BasicAWSCredentials(
+                  s3ConfigProperties.getAccessKeyId().getPassword(),
+                  s3ConfigProperties.getSecretAccessKey().getPassword());
+              amazonS3ClientBuilder.withCredentials(new AWSStaticCredentialsProvider(creds));
+            }
+            return new ServerSideEncryptingAmazonS3(amazonS3ClientBuilder.build(), storageConfig.getServerSideEncryption());
+          } else {
+            return Preconditions.checkNotNull(s3Client, "s3Client");
```

Comment:
hmm, should the check that one of `s3Client` or `amazonS3ClientBuilder` and `storageConfig ` are not null be done eagerly instead of in the supplier? (i'm not certain either way, just thinking out loud)

Reply:
I think it will be the same either way. Had it this way to keep all relevant logic for determining which s3Client to use together (inside the Suppliers.memorize(). I'm fine either way though.

Reply:
I guess it may not matter much here since this is just a sanity check which shouldn't happen, but it would be better to fail early as possible. The check in the constructor will be called in the JSON deserialization while other methods will be called after the task is scheduled on a middleManager.

Reply:
Moved all the Preconditions.checkNotNull outside